src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs

```diff
@@ -16,43 +16,268 @@ namespace Mono.Linker.Steps
+		/// It may optimize other methods as well - those are remembered for future reuse.
+		/// </summary>
+		/// <param name="method">The method to process</param>
+		void ProcessMethod (MethodDefinition method)
+		{
+			Debug.Assert (processingStack.Count == 0 && processingMethods.Count == 0);
+			processingStackVersion = 0;
+
+			if (!processedMethods.ContainsKey (method)) {
+				AddMethodForProcessing (method);
+
+				ProcessStack ();
+			}
+
+			Debug.Assert (processedMethods.ContainsKey (method));
+		}
+
+		void AddMethodForProcessing (MethodDefinition method)
+		{
+			Debug.Assert (!processedMethods.ContainsKey (method));
+
+			var processingNode = new ProcessingNode (method, -1);
+
+			var stackNode = processingStack.AddFirst (processingNode);
+			processingMethods.Add (method, stackNode);
+			processingStackVersion++;
+		}
+
+		void StoreMethodAsProcessedAndRemoveFromQueue (LinkedListNode<ProcessingNode> stackNode, Instruction methodValue)
+		{
+			Debug.Assert (stackNode.List == processingStack);
+			Debug.Assert (methodValue != null);
+
+			var method = stackNode.Value.Method;
+			processingMethods.Remove (method);
+			processingStack.Remove (stackNode);
+			processingStackVersion++;
+
+			processedMethods[method] = methodValue;
 		}
 
-		bool TryGetConstantResultInstructionForMethod (MethodDefinition method, out Instruction constantResultInstruction)
+		void ProcessStack ()
 		{
-			if (constExprMethods.TryGetValue (method, out constantResultInstruction))
-				return constantResultInstruction != null;
+			while (processingStack.Count > 0) {
+				var stackNode = processingStack.First;
+				var method = stackNode.Value.Method;
+
+				bool treatUnprocessedDependenciesAsNonConst = false;
+				if (stackNode.Value.LastAttemptStackVersion == processingStackVersion) {
+					// Loop was detected - the stack hasn't changed since the last time we tried to process this method
+					// as such there's no way to resolve the situation (running the code below would produce the exact same result).
+
+					// Observation:
+					//   All nodes on the stack which have `LastAttemptStackVersion` equal to `processingStackVersion` are part of the loop
+					//   meaning removing any of them should break the loop and allow to make progress.
+					//   There might be other methods in between these which don't have the current version but are dependencies of some of the method
+					//   in the loop.
+					//   If we don't process these, then we might miss constants and branches to remove. See the doc
+					//   `constant-propagation-and-branch-removal.md` in this repo for more details and a sample.
+
+					// To fix this go over the stack and find the "oldest" node with the current version - the "oldest" node which
+					// is part of the loop:
+					var lastNodeWithCurrentVersion = stackNode;
+					for (var currentNode = stackNode; currentNode != null; currentNode = currentNode.Next) {
```

Comment:
This will walk the whole stack . I'm not sure it matters since we expect loop cases are rare, but would lead to bad worst-case performance. A small optimization would be to walk the list backwards.

Reply:
That's a good idea - changed.